main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -175,15 +187,38 @@ object Lexer {
     val c = s.src.data(s.current.offset)
     if (c == '\n') {
       s.current.line -= 1
-      s.current.column = 0
-      s.end.line -= 1
-      s.end.column = 0
+      s.current.column = s.prevWidth
+      recomputePrevWidth()
     } else {
       s.current.column -= 1
-      s.end.column -= 1
     }
   }
 
+  /**
+    * Assuming that `s.current` is on the end of a line (i.e. on `\n`), recomputes the length of
+    * the previous line.
+    */
+  private def recomputePrevWidth()(implicit s: State): Unit = {
```

Comment:
Will this be slow?

Reply:
It traverses a line, so somewhat. but 1) there is a (a little) less overhead during normal advance nad 2) I think we should delete retreat and do it with checkpoints, then this function will also be deleted 3) it is only triggered if you retreat across a line 4) it is stored in state, so its not necessarily spending effort. 5) i could do it on-demand if needed, but due to point 2 i don't think its worth it